src/uct/ib/rc/accel/rc_mlx5_ep.c

```diff
@@ -968,80 +968,87 @@ UCS_CLASS_INIT_FUNC(uct_rc_mlx5_ep_t, const uct_ep_params_t *params)
     return status;
 }
 
-static void uct_rc_mlx5_ep_clean_qp(uct_rc_mlx5_ep_t *ep, uct_ib_mlx5_qp_t *qp)
+typedef struct {
+    uct_rc_ep_cleanup_ctx_t    super;
+    uct_rc_mlx5_iface_common_t *iface;
+    uct_ib_mlx5_qp_t           tm_qp;
+    uct_ib_mlx5_qp_t           qp;
+    uct_ib_mlx5_mmio_reg_t     *reg;
+} uct_rc_mlx5_ep_cleanup_ctx_t;
+
+static int uct_rc_mlx5_ep_clean_filter(const ucs_callbackq_elem_t *elem,
+                                       void *arg)
 {
-    uct_rc_mlx5_iface_common_t *iface = ucs_derived_of(ep->super.super.super.iface,
-                                                       uct_rc_mlx5_iface_common_t);
+    return (elem->cb == (void *)uct_rc_mlx5_ep_destroy) && (elem->arg == arg);
+}
+
+void uct_rc_mlx5_ep_destroy(uct_rc_ep_cleanup_ctx_t *rc_ep_cleanup_ctx)
+{
+    uct_rc_mlx5_ep_cleanup_ctx_t *ep_cleanup_ctx
+                                      = ucs_derived_of(rc_ep_cleanup_ctx,
+                                                       uct_rc_mlx5_ep_cleanup_ctx_t);
+    uct_rc_mlx5_iface_common_t *iface = ep_cleanup_ctx->iface;
     uct_ib_mlx5_md_t *md              = ucs_derived_of(iface->super.super.super.md,
                                                        uct_ib_mlx5_md_t);
 
-    /* Make the HW generate CQEs for all in-progress SRQ receives from the QP,
-     * so we clean them all before ibv_modify_qp() can see them.
-     */
-#if HAVE_DECL_IBV_CMD_MODIFY_QP && !HAVE_DEVX
-    struct ibv_qp_attr qp_attr;
-    struct ibv_modify_qp cmd;
-    int ret;
-
-    /* Bypass mlx5 driver, and go directly to command interface, to avoid
-     * cleaning the CQ in mlx5 driver
-     */
-    memset(&qp_attr, 0, sizeof(qp_attr));
-    qp_attr.qp_state = IBV_QPS_RESET;
-    ret = ibv_cmd_modify_qp(qp->verbs.qp, &qp_attr, IBV_QP_STATE, &cmd, sizeof(cmd));
-    if (ret) {
-        ucs_warn("modify qp 0x%x to RESET failed: %m", qp->qp_num);
+    uct_ib_device_async_event_unregister(&md->super.dev,
+                                         IBV_EVENT_QP_LAST_WQE_REACHED,
+                                         ep_cleanup_ctx->qp.qp_num);
+    ucs_callbackq_remove_if(&iface->super.super.super.worker->super.progress_q,
+                            uct_rc_mlx5_ep_clean_filter, ep_cleanup_ctx);
+    ucs_list_del(&ep_cleanup_ctx->super.list);
+    uct_rc_mlx5_iface_common_check_cqs_ci(iface, &iface->super.super);
+
+#if IBV_HW_TM
+    if (UCT_RC_MLX5_TM_ENABLED(iface)) {
+        uct_ib_mlx5_iface_put_res_domain(&ep_cleanup_ctx->tm_qp);
```

Comment:
whye it's needed for tm_qp but not for regular qp?

Reply:
check if we can use mmio_cleanup here too

Reply:
comment why using uct_ib_mlx5_iface_put_res_domain and not uct_ib_mlx5_qp_mmio_cleanup. because: in case of devx, we don't have uar, and uct_ib_mlx5_qp_mmio_cleanup would try to release uar